### ISSUE TITLE
app not remembering answer for rooted device

### ISSUE THREAD

**Issue by klic** [corona-warn-app/cwa-app-android#4264]
## Avoid duplicates
* [x] Bug is not mentioned in the [FAQ](https://www.coronawarn.app/en/faq/)
* [x ] Bug is specific for Android only, for general issues / questions that apply to iOS and Android please raise them in the [documentation repository](https://github.com/corona-warn-app/cwa-documentation)
* [ x] Bug is not already reported in another issue

## Technical details

- Fairphone FP3
- Android version:LOS 18
- App version:2.12.1

## Describe the bug

App warns that the phone is rooted, which is fine by me. But the app does not remember the answer and shows the same pop-up again with every start. 

## Steps to reproduce the issue
Start the app

## Expected behaviour

The app should remember your answer, and only ask again if something changed

## Additional context

---
Internal Tracking ID: [EXPOSUREAPP-10181](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-10181)
Internal Tracking ID: [EXPOSUREAPP-10698](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-10698)

**Comment by chr4** [comment]
Another hint: This is also displayed as a false-positive on non-rooted devices with ROMs such as LineageOS.

(It feels like the "root warnings" come a little out of hand and seem to be included in some sort of "compliance" rule somewhere. If you care for security, maybe consider checking Android security patchlevel instead. This nags people with insecure devices instead of tech-savvy people with custom ROMs, often installed for security and privacy reasons).

**Comment by MikeMcC399** [comment]
I also found the root warning being triggered on an Android emulator (8" Fold-out using API 30) which I was using to look at layout issues on newer larger smartphones - see https://github.com/corona-warn-app/cwa-app-android/issues/4203. This admittedly is not mainstream usage and, for me, is only an annoyance for testing.

**Comment by scoder** [comment]
I checked the ADB logs and the thing that triggers the warning for me is "ro.debuggable" in rootbeer. So … my device goes as debuggable *iff* I enable debugging? That's not very surprising. And probably ok to show a warning in that case, because, ok, someone's eves-dropping on the logs.

But it counts as "debuggable" even when debugging is *disabled*? And it shows a warning for that? That counts as a false positive for me. And an annoying one.

Now, the question is, what's wrong here? The fact that CWA shows a warning? (at all? in this case?) The fact that rootbeer considers this an issue? Or the fact that the "ro.debuggable" property is set in the first place?

I don't know.

**Comment by thomasaugsten** [comment]
Can you check with root beer sample app which check is marked for you

**Comment by dsarkar** [comment]
@scoder See https://github.com/corona-warn-app/cwa-app-android/issues/4264#issuecomment-956097442

**Comment by scoder** [comment]
CWA logs it as `ro.debuggable` (before enabling ADB logging), so I'd say that's it.

**Comment by x2k13** [comment]
https://github.com/corona-warn-app/cwa-app-android/issues/4268#issuecomment-961805351

**Comment by vaubaehn** [comment]
@dsarkar @heinezen 
If BSI/BMG/RKI really insist of having a warning for those devices, I propose to change the way to "inform" user:
- The dialog that the device is rooted (or has an opened boot loader) should be shown only once, and never again.
- There could be a hint in the home screen, either as an informational tile like the incompatibility warning (https://github.com/corona-warn-app/cwa-app-android/issues/3170) or, if you like a different design, as a simple/colored text ("insecure device") below the tracing status switch.
For me looks like best compromise to fulfill the requirement of authorities and to not nag advanced users more than necessary.
What do you think?

**Comment by x2k13** [comment]
> @dsarkar @heinezen If BSI/BMG/RKI really insist on having a _permanent_ warning for those devices, I propose to change the way to "inform" user:
> 
> * The dialog that the device is rooted (or has an opened boot loader) should be shown only once, and never again.
> * There could be a hint in the home screen, either as an informational tile like the incompatibility warning ([Incompatibility warning after updating to version 2.2.1 #3170](https://github.com/corona-warn-app/cwa-app-android/issues/3170)) or, if you like a different design, as a simple/colored text ("insecure device") below the tracing status switch.
> 
> For me looks like best compromise to fulfill the requirement of authorities and to not nag advanced users more than necessary. What do you think?

I don't like the idea of a permanent warning banner or whatever in a prominent position in the app (read: in your face).
This would make a lot of people frown and wonder, especially if the device isn't rooted at all.

Better change to a library/piece of code that correctly checks root or no root and doesn't claim a device to be unsafe if it in fact isn't.

**Comment by vaubaehn** [comment]
Hi @x2k13 ,
thanks for your feedback.

We need to separate two different issues:
1. rootbeer reporting false positives
2. the way of displaying a warning in case of a true positive.

My proposal above was for point 2. only: displaying a warning in case of a true positive. It looks like it's a requirement/strong recommendation from authorities, to have a warning displayed when the device is rooted. The way to achive this that I proposed might be the least annoying way. Advanced users that intentionally rooted their devices will continue to ignore the warning. In the rare cases where unadvanced users have an unrooted device by accident/by abuse from third parties, they will seek help to understand, what is going on - and can then decide on what to do next.

> I don't like the idea of a permanent warning banner or whatever in a prominent position in the app (read: in your face).
> This would make a lot of people frown and wonder, especially if the device isn't rooted at all.
> Better change to a library/piece of code that correctly checks root or no root and doesn't claim a device to be unsafe if it in fact isn't. Banking apps are known to be really picky about rooted devices, but these recognise root more efficiently and correctly than CWA does atm.

Related to the false positives, point 1.: It will need to be determined more clearly, what actually is a false positive: A mainstream device from a known brand with stock rom and no root which is detected of having root falsely? Or a customized mobile with an open bootloader but no root - could the open boot loader be abused? When there is a common agreement, and there is actually a bug in the root detector, it should be fixed.
In either way, even in these cases of false positives, it may be better to have a less annoying display of a warning instead of a multiple-times-pop-up of a warning dialog. Won't you agree?

**Comment by darkmattercoder** [comment]
> It looks like it's a requirement/strong recommendation from authorities, to have a warning displayed when the device is rooted. 

Do you have a reference link to this _requirement or strong recommendation_? Things do not necessarily make sense just because everyone else is doing it.
Is there any BSI notice where this is clearly stated? I doubt it.

> The way to achive this that I proposed might be the least annoying way. Advanced users that intentionally rooted their devices will continue to ignore the warning. In the rare cases where unadvanced users have an unrooted device by accident/by abuse from third parties, they will seek help to understand, what is going on - and can then decide on what to do next.

And when (If that will ever happen), someone has to scan your certificate in your CWA, sees a permanent red banner and does not accept the certificate and rejects your attendance at an event, just because "oh you obviously hacked the cwa app to trick me"?

A permanent banner, exactly as an eacht-time warning on startup _is_ and _stays_ an act of paternalism.

Why to care about false positives in the first place? If anyone gets a false positive warning which carries a checkbox that will lead to a suppression of subsequent warnings, they can acknowledge it and do not get bothered anymore.

**Comment by vaubaehn** [comment]
@darkmattercoder 
> Do you have a reference link to this _requirement or strong recommendation_? Things do not necessarily make sense just because everyone else is doing it.
> Is there any BSI notice where this is clearly stated? I doubt it.

That introducing a "root warning" may be a requirement from authorities is my assumption based on these experiences:
- BSI takes care a lot of issues, that are not always considered to be a threat neither by (advanced) users nor by our developers. Just take a look here, read a bit, and then it should be obvious, that at least it is rather likely that BSI actively suggested to introduce this "feature": https://github.com/corona-warn-app/cwa-app-android/issues?q=+is%3Aissue+author%3Absi-tf-cwa That we didn't see an open issue created by @BSI-TF-CWA here to implmement a "root warning" anymore could be the result of recent complaints of SAP staff and also of users, not to disclose security related issues via github anymore.
- I'm following here for quite a while, and just by discussions with devs and project leads I don't have the feeling that they were keen to introduce that "feature" on their own action. I guess the devs here share many of your ideas/opinions/attitudes - but for some good reason they can not always shout out loud how happy they are to introduce features like the "root warning" 😉 

So, partly you're true, I don't have a proof that there was a requirement from "outside"/authorities to implement a root warning, even it's likely they did. Maybe @thomasaugsten or @mlenkeit or even the @BSI-TF-CWA could share some details who wanted to have it integrated and why?

> And when (If that will ever happen), someone has to scan your certificate in your CWA, sees a permanent red banner and does not accept the certificate and rejects your attendance at an event, just because "oh you obviously hacked the cwa app to trick me"?

That is indeed a **_very_** good point that needs to be taken into account! At least, after you activated the app and switched to the "certificate screen", no warning is/should be visible.

> A permanent banner, exactly as an eacht-time warning on startup _is_ and _stays_ an act of paternalism.

This is a political discussion. My feeling is, that the person who recommended to implement this feature actually wanted to protect users who have no idea that their mobile is "vulnerable" and their CWA data (ie. medical data) is unsafe. This person judged the protection of clueless users higher than nagging advanced users with intentionally customized OS. Please don't discuss this special point with _me_, you would run into an open door. Generally speaking, who feels paternalized and when is a big social and cultural issue currently. How many people feel paternalized because they need to prove their vaccination/test/recovery status when they want to enter a bar? I have not root, currently. But if I had, I'd be happy to be able to simply ignore the warning and I would feel good that other clueless users with unintentional root are now protected. Why should I care more? It's not to nag me but protect others.

> Why to care about false positives in the first place? If anyone gets a false positive warning which carries a checkbox that will lead to a suppression of subsequent warnings, they can acknowledge it and do not get bothered anymore.

That has been discussed here before. I don't know why this hasn't been implemented like this. Either there is the _requirement_ to have a permanent warning, or a change of behavior was not developing priority yet. Maybe @dsarkar or @heinezen could bring in some light here?

In any case, enjoy your week-end!

**Comment by vaubaehn** [comment]
@x2k13 
The reason why the root beer lib signaled a red alert, could be the DangerousProps:
If you feel nagged, you may look here https://www.reddit.com/r/LineageOS/comments/cyy9aj/how_to_get_pass_on_native_root_checker/ and follow the instructions from the first reply.
Basically, it's editing the `default.prop` and change some parameters to be more "secure", like explained in the introduction here: https://blog.snapdragon.cc/2019/09/03/fully-unroot-custom-android-roms-to-run-banking-apps/. If you're able to directly edit the props, that alone might be reason enough to give a warning...

**Comment by x2k13** [comment]
> @x2k13 The reason why the root beer lib signaled a red alert, could be the DangerousProps: If you feel nagged, you may look here https://www.reddit.com/r/LineageOS/comments/cyy9aj/how_to_get_pass_on_native_root_checker/ and follow the instructions from the first reply. Basically, it's editing the `default.prop` and change some parameters to be more "secure", like explained in the introduction here: https://blog.snapdragon.cc/2019/09/03/fully-unroot-custom-android-roms-to-run-banking-apps/. If you're able to directly edit the props, that alone might be reason enough to give a warning...

Yes, you guessed right. I already wrote about that here before:
https://github.com/corona-warn-app/cwa-app-android/issues/4264#issuecomment-961808532

**Comment by vaubaehn** [comment]
> 
> 
> CWA logs it as `ro.debuggable` (before enabling ADB logging), so I'd say that's it. EDIT: tried the sample app, just in case, same result: dangerous props.

@scoder please see https://github.com/corona-warn-app/cwa-app-android/issues/4264#issuecomment-962196355
Though there is probably not much you can do, except rooting your phone, changing `default.prop` params to safe values and unrooting again.

**Comment by klic** [comment]
This is an interesting discussion, imho. 

What I again and again am flabbergasted is, postmarket and current, rooted, roms are flagged, but people with lollipop roms go unmolested. Sure, they shouldn't been forced to throw away their devices, but the whole thing is a total hybris. People who try to stay safe are flagged, and people, who are unsafe at a probably high degree, and maybe don't know it, are lulled in safety. That's absurd. 
Of course, they can show me ONCE, that i don't have a stock rom, or am rooted, or whatnot. But if i did that deliberately, I acknowledge it, because i know. And if I'm clueless, i probably acknowledge it also, because it's way above my head (and honestly, those people are not to blame, this is consumer electronics, you should not need a cs degree for it! ). So in the end, it's snake oil. And absolute helplessness of any agency, be it a BSI or whatever. That's not honest

**Comment by darkmattercoder** [comment]
As this ist an Open source project and the topic has been around for a while, I wonder what the maintainer's perspective is and what their discussions internally are. We need a transparent information flow! @devs could you please elaborate and share your plans regarding this?

**Comment by darkmattercoder** [comment]
Still not worth elaborating on this topic from an official? This project is funded trough public money. The code already is public also. It would be nice to also have publicly available answers to upcoming questions.

I think that for now, one should be better off by using CCTG instead of CWA. It is apparently the better choice for people who are aware of their rooted devices.

**Comment by mhier** [comment]
Fun fact: With Magisk Hide enabled, the warning will disappear. This shows how stupid the idea behind such messages is: People who intentionally root their phones (or not even that, just install LineageOS or such) already know what they have done. They do not need to see this message. If anyone with bad intentions somehow sneaks some root back door onto a phone, it will likely not be detected by other apps and hence no warning will appear.

I wonder how secure an app can be if its developers have not understood such basic principle...

**Comment by mlenkeit** [comment]
> Fun fact: With Magisk Hide enabled, the warning will disappear.

No worries, we are well aware of these limitations, and we also made this transparent to all involved parties 😉

**Comment by 2019-05-10** [comment]
> Still not worth elaborating on this topic from an official?

Given it takes almost 6 months to add a few words to the FAQ, you cannot expect an answer to so complex a question in a matter of days!
Nevertheless, I came here because I was annoyed by that pointless nagging as well and intended to file a bug about it.
So, I am very much interested about the reasoning behind that user-unfriendly and rather meaningless decision (as several people already pointed out) as well.

**Comment by dsarkar** [comment]
Internal Tracking ID: [EXPOSUREAPP-10181](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-10181)
Internal Tracking ID: [EXPOSUREAPP-10698](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-10698)

**Comment by dsarkar** [comment]
Dear @2019-05-10, dear community, 

Obviously this feature was implemented not to "nag" anybody, but because stakeholders requested so, and this is related to the strong protection of privacy, that might be undermined on rooted devices (of course not necessarily). 

The main thought was, that is my understanding, that some users might not know that their device is rooted and therefore there might exist some possibility of vulnerability. To answer the question, why would somebody not know about his device being rooted: Second hand device, somebody gets a new phone and gives away his rooted device,...

However, due to the large feedback, it is being discussed and planned to change the behaviour, i.e. the frequency how often this message shall be shown and in which situation.

Of course we understand that, in particular, the more technically inclined user feels somehow "annoyed", however, it is, as usual, always quite a balancing act to find the middle ground between all the different requirements and criteria.

Anyway, as said, we might expect some changes.

Best wished, DS

---
Corona-Warn-App Open Source Team

**Comment by mhier** [comment]
@dsarkar I am sorry, but in my opinion this is a very constructed situation. It should be very obvious that this message hits almost exclusively people which are very well aware that their device is rooted. On the other hand even the concern about second hand devices is also misguided: rooting itself does *not* compromise privacy in any way. Only if apps are installed which intentionally compromise privacy this poses a problem. Now the key point is, especially for second hand devices it is very likely that no rooting is required for this. Any older Android version (as any other older operating system) has security holes which can be abused.

So if stakeholders were really concerned about about privacy, why does a rooted device give a warning but the app can run on Android 6.0 without any warning? In my eyes this is not plausible. Either the stakeholders have no sense about software security, or they want to nag us. Both ways this is bad.

You might say I am overreacting. Yes maybe, but this is not the first app annoying people like me with such messages. This is especially problematic if apps are not only showing annoying messages but stop functioning because of such misguided arguments. So please, dear Corona-Warn-App Open Source Team, take this message back to your stakeholders. This is not protecting privacy, it is not making any thing more secure. The only effect this kind of false protection has is to make devices quicker obsolete (because people get pushed away from using aftermarket firmware with longer support). This stupid message has an impact on the environment! So please stop it!

**Comment by 2019-05-10** [comment]
> the stakeholders have no sense about software security

This one.
Remember, those mysterious stakeholders are the same people responsible for monoculture of MS products in the public administration.
Tells you everything you need to know about their idea of "security" (I guess, they are used to simply dismiss any popup anyway w/o reading it).

**Comment by Ein-Tim** [comment]
PR:  #4550

**Comment by dsarkar** [comment]
Hi @klic,
Thanks for your suggestion. We suggest that with the implementation of R 2.16 this issue can be closed.

Best wishes, DS

---
Corona-Warn-App Open Source Team

**Comment by mhier** [comment]
I still think showing this message at all is wrong. The app should not even perform any such check. It's not the app's business whether it runs on a rooted phone. Please remove any SafetyNet check from the app. This app is not even remotely security related. Stop poking your nose into such business.

**Comment by vaubaehn** [comment]
Hi all guys with a rooted phone, we could need your help in a different issue.
There may be a bug in the Exposure Notification System that leads to filling up internal storage until the system runs out of free space.
If you may help, please leave a note over https://github.com/corona-warn-app/cwa-app-android/issues/4732
Thank you!

**Comment by Ein-Tim** [comment]
@dsarkar Please close this issue now.

**Comment by Ein-Tim** [comment]
@mhier

> The app should not even perform any such check. It's not the app's business whether it runs on a rooted phone. 

I suggest to [open a new enhancement request](https://github.com/corona-warn-app/cwa-app-android/issues/new?assignees=&labels=enhancement&template=02_enhancement.md) reg. this.

> This app is not even remotely security related. 

While it is not security related, it stores highly sensitive health information, including information on vaccination, recovery & test.

**Comment by MikeMcC399** [comment]
I support closing this issue. The original problem description is solved.

I verified on current versions. Although it is a false positive to declare a Pixel 3a emulator running Android 6.0 as a rooted device, I used it to check the suppression of the message until an update was done.

1. Install CWA 2.20.4 on Pixel 3a emulator API 23 (Android 6.0)
2. "Security Note" regarding root is immediately displayed
3. Tap "OK"
4. Complete onboarding
5. FORCE STOP CWA
6. Open CWA and confirm that "Security Note" is shown ![image](https://user-images.githubusercontent.com/66998419/164758958-27271a97-95ec-433f-8e96-93950c6ca659.png)
7. Tap on "Hide notification until next update" to enable, click OK
8. FORCE STOP CWA
9. Open CWA and confirm that **no** "Security Note" is shown
10. Update CWA to 2.21.1
11. Confirm that "Security Note" is shown again
12. Tap on "Hide notification until next update" to enable, click OK
13. Tap "Next" to acknowledge new version information
14. FORCE STOP CWA
15. Open CWA and confirm that **no** "Security Note" is shown
